### ISSUE TITLE
Commit loss is negative

### ISSUE THREAD

**Issue by ZekaiGalaxy** [lucidrains/meshgpt-pytorch#43]
<img width="1224" alt="image" src="https://github.com/lucidrains/meshgpt-pytorch/assets/77915605/dcaa662f-cb28-4038-a5cb-ccdbf1b99080">

When I trained on several objects with several epochs, the commit loss starts to become negative, and it turns out that the overall loss keeps going down, but neither the recon loss nor the reconstruction result turns better. 

I wonder if the commit loss being negative is normal or not, or what it implies

**Comment by MarcusLoppe** [comment]
Try lowering the diversity_gamma from 1.0 to 0.1 - 0.3.
The quantizer uses this variable to lower the loss, so it uses 1.0 which allows it to trick itself to be more diverse since it punishes exploration less.  As you can see,, your commit loss is near -1 which is due to the code below, I think it's good to have high diversity at start but at the end it might do more harm then good.

Part of the commit loss calculation in LFQ:
`entropy_aux_loss = per_sample_entropy - self.diversity_gamma * codebook_entropy`

 
```
autoencoder = MeshAutoencoder(    
    num_discrete_coors = 128 ,
    rlfq_kwargs = {"diversity_gamma": 0.2 }
)
```

**Comment by qixuema** [comment]
> Try lowering the diversity_gamma from 1.0 to 0.1 - 0.3. The quantizer uses this variable to lower the loss, so it uses 1.0 which allows it to trick itself to be more diverse since it punishes exploration less. As you can see,, your commit loss is near -1 which is due to the code below, I think it's good to have high diversity at start but at the end it might do more harm then good.
> 
> Part of the commit loss calculation in LFQ: `entropy_aux_loss = per_sample_entropy - self.diversity_gamma * codebook_entropy`
> 
> ```
> autoencoder = MeshAutoencoder(    
>     num_discrete_coors = 128 ,
>     rlfq_kwargs = {"diversity_gamma": 0.2 }
> )
> ```
Hi @MarcusLoppe 

Happy New Year! 🎉🎉🎉

I attempted to train an autoencoder using 20 different chairs as training samples and encountered the same issue where the commit loss was negative. 

This is the [commit loss curve](https://wandb.ai/xueqi/mesh_ae/reports/commit_loss-23-12-28-23-07-44---Vmlldzo2MzY5NTY4) during my training process. 

![W B Chart 2023_12_28 23_14_01](https://github.com/lucidrains/meshgpt-pytorch/assets/58554118/cb5cf2d4-cfc2-41f0-85e3-42ac6ecf3d14)

I will reduce the diversity_gamma from 1.0 to between 0.1 and 0.3 to see what changes occur in the commit loss.

Best regards,
Xueqi Ma

**Comment by MarcusLoppe** [comment]
> > Try lowering the diversity_gamma from 1.0 to 0.1 - 0.3. The quantizer uses this variable to lower the loss, so it uses 1.0 which allows it to trick itself to be more diverse since it punishes exploration less. As you can see,, your commit loss is near -1 which is due to the code below, I think it's good to have high diversity at start but at the end it might do more harm then good.
> > Part of the commit loss calculation in LFQ: `entropy_aux_loss = per_sample_entropy - self.diversity_gamma * codebook_entropy`
> > ```
> > autoencoder = MeshAutoencoder(    
> >     num_discrete_coors = 128 ,
> >     rlfq_kwargs = {"diversity_gamma": 0.2 }
> > )
> > ```
> 
> Hi @MarcusLoppe
> 
> Happy New Year! 🎉🎉🎉
> 
> I attempted to train an autoencoder using 20 different chairs as training samples and encountered the same issue where the commit loss was negative.
> 
> This is the [commit loss curve](https://wandb.ai/xueqi/mesh_ae/reports/commit_loss-23-12-28-23-07-44---Vmlldzo2MzY5NTY4) during my training process.
> 
> ![W B Chart 2023_12_28 23_14_01](https://private-user-images.githubusercontent.com/58554118/293230219-cb5cf2d4-cfc2-41f0-85e3-42ac6ecf3d14.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM3Nzg3NTksIm5iZiI6MTcwMzc3ODQ1OSwicGF0aCI6Ii81ODU1NDExOC8yOTMyMzAyMTktY2I1Y2YyZDQtY2ZjMi00MWYwLTg1ZTMtNDJhYzZlY2YzZDE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI4VDE1NDczOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJkMGFhMDA0NjA0YTNkZWMzYTUwZTg2NjI2YmUyNjQ4OWYxMDJlMjJlY2U1MzEwM2MyOWJhM2M4YmNjMWZjOTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.U1DGTzfcNf2-rhOWnHCtdFAgZMY6SM3REw_0EK01FtE)
> 
> I will reduce the diversity_gamma from 1.0 to between 0.1 and 0.3 to see what changes occur in the commit loss.
> 
> Best regards, Xueqi Ma

Experiment a little bit since I just discovered this yesterday and haven't tested it out fully :)
Diversity is good at the start of the training but changing too much at the end of the training isn't very good.

Please let me know what you find out.

**Comment by qixuema** [comment]
@MarcusLoppe 

I tried reducing the diversity_gamma from 1.0 to 0.2 and retrained the data. The current [commit loss curve](https://wandb.ai/xueqi/mesh_ae/reports/commit_loss-23-12-29-09-05-56---Vmlldzo2MzcyMDky) is shown in the following image.

![W B Chart 2023_12_29 09_05_36](https://github.com/lucidrains/meshgpt-pytorch/assets/58554118/2c987903-b069-4212-bc7c-e7c3268365d7)

**Comment by ZekaiGalaxy** [comment]
Thank you @MarcusLoppe, From my perspective, maybe we can use a 'decaying' gamma, since we want it to explore at the beginning but converge at the end. 

I also notice that in @qixuema 's experiment, with gamma = 0.2  commit loss do drop, but  there are some extreme commit loss values. Does that mean the model overfits to a certain or several types of shapes, codes, whatever and can't do rare cases well.

**Comment by ZekaiGalaxy** [comment]
Also @qixuema how's your recon loss going? I find that though my recon loss is going down (~0.32), I still can't reconstruct the train data using autoencoder when trained on multi objects.

**Comment by qixuema** [comment]
Hi, @ZekaiGalaxy 

The following are my [recon_loss](https://wandb.ai/xueqi/mesh_ae/reports/recon_loss-23-12-29-14-59-58---Vmlldzo2MzczNzk2) and [total_loss](https://wandb.ai/xueqi/mesh_ae/reports/total_loss-23-12-29-15-00-12---Vmlldzo2MzczODAx).

![W B Chart 2023_12_29 14_51_00](https://github.com/lucidrains/meshgpt-pytorch/assets/58554118/a7c492d7-dd5a-44e0-872b-c5dbfc437fbd)

![W B Chart 2023_12_29 14_50_48](https://github.com/lucidrains/meshgpt-pytorch/assets/58554118/d9373e4d-0c43-41c1-916b-80df420845c2)

**Comment by MarcusLoppe** [comment]
@lucidrains  


Hi all, this issue resolves itself when training on a large dataset, using a 300x50 augmentations the commit was at 3-14 at start and then settled itself and matched the recon loss at around  0.6

**Comment by avramdj** [comment]
Hey, isn't this still a problem with the commit loss implementation itself? Any plans on fixing it?
